packages/tupaia-web/src/features/DashboardItem/DashboardItemContent.tsx

```diff
@@ -53,30 +60,62 @@ const LoadingContainer = styled.div`
 
 const DisplayComponents = {
   chart: Chart,
+  view: View,
   matrix: Matrix,
 };
 
 interface DashboardItemContentProps {
-  viewContent: DashboardItemType & (ChartViewContent | MatrixViewContent);
+  config: DashboardItemType;
+  report: DashboardItemReport;
   isEnlarged?: boolean;
   isLoading: boolean;
   error: UseQueryResult['error'] | null;
   onRetryFetch: UseQueryResult['refetch'];
   isExpandable: boolean;
 }
 
+export const getNoDataString = (report: DashboardItemReport, config: DashboardItemType) => {
+  const { startDate, endDate } = report;
+  const { noDataMessage, source } = config;
+  if (noDataMessage) {
+    return noDataMessage;
+  }
+
+  if (source === 'mSupply') {
+    return 'Requires mSupply';
+  }
+
+  if (startDate && endDate) {
+    return `No data for ${startDate} to ${endDate}`;
+  }
+
+  return 'No data for selected dates';
+};
+
+const getShowNoDataMessage = (report: DashboardItemReport, type: DashboardItemConfig['type']) => {
```

Comment:
```suggestion
const getHasNoData = (report: DashboardItemReport, type: DashboardItemConfig['type']) => {
```
It would be nice if this name was a bit more declarative rather than imperative.